File: cpp/word-distance.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>
#include <cassert>
#include "word-distance.h"
#include "mtree.h"
using namespace std;


//const char DICT_FILE[] = "pt-br.dic";
const char DICT_FILE[] = "en.dic";



int main(int argc, const char* argv[]) {
	//cout << wordDistance("gol", "bola") << endl;
	//cout << wordDistance(argv[1], argv[2]); return 0;

	size_t wordsLimit = (argc > 1) ? atoi(argv[1]) : 1000000;

	WordMTree mtree;

	size_t loadedWords = 0;
	Timer t;
	cout << "Indexing..." << flush;
	ifstream f(DICT_FILE);
	while(! f.eof()) {
		string line;
		getline(f, line);

		if(!line.empty()  &&  line[0] != '%') {
			//word = unicode(line.strip(), 'utf-8')	TODO: implement encoding
			auto word = line;
			mtree.add(word);
			loadedWords++;
			if(loadedWords >= wordsLimit) {
				break;
			}
			if(loadedWords % 100 == 0) {
				cout << '\r' << loadedWords << " words indexed" << flush;
			}
		}
	}
	cout << '\r' << loadedWords << " words indexed" << endl;

	Timer::Times times = t.getTimes();
	printf("TIMES: %0.2fuser %0.02fsys %0.2freal\n\n", times.user, times.sys, times.real);

	while(true) {
		//word = unicode(raw_input("Type a word: "), 'utf-8')	TODO: implement encoding
		cout << "Type a word: ";
		string word;
		getline(cin, word);
		if(word == "") {
			break;
		}

		Timer t;
		auto query = mtree.get_nearest_by_limit(word, 10);
		for(auto i = query.begin(); i != query.end(); ++i) {
			cout << "\t" << i->distance << " " << i->data << endl;
		}
		Timer::Times times = t.getTimes();
		printf("TIMES: %0.2fuser %0.02fsys %0.2freal\n\n", times.user, times.sys, times.real);
	}
}

File: py/word_distance.py
#!/usr/bin/env python
import sys
import time
from mtree import MTree


#DICT_FILE = 'pt-br.dic'
DICT_FILE = 'en.dic'


class Timer(object):
	def __init__(self):
		self.clock = time.clock()
		self.time = time.time()
	
	def getTimes(self):
		c = time.clock()
		t = time.time()
		return (c - self.clock, t - self.time)



def word_distance(word1, word2):
	word1 = word1.lower()
	word2 = word2.lower()
	
	
	distance = [
		[None for _ in range(len(word2)+1)]
			for _ in range(len(word1)+1)
	]
	
	distance[0][0] = 0
	for len1 in range(1, len(word1)+1):
		distance[len1][0] = len1
	for len2 in range(1, len(word2)+1):
		distance[0][len2] = len2
	
	for len1 in range(1, len(word1)+1):
		for len2 in range(1, len(word2)+1):
			if word1[len1-1] == word2[len2-1]:
				distance[len1][len2] = distance[len1-1][len2-1]
			else:
				distance[len1][len2] = 1 + min([
						distance[len1-1][len2-1],
						distance[len1  ][len2-1],
						distance[len1-1][len2  ],
				])
	
	return distance[len(word1)][len(word2)]



def main():
	#print word_distance('gol', 'bola')
	#print word_distance(*sys.argv[1:])
	
	words_limit = int(sys.argv[1])
	
	mtree = MTree(distance_function=word_distance)
	
	loaded_words = 0
	t = Timer()
	print 'Indexing...', ; sys.stdout.flush()
	with open(DICT_FILE) as f:
		for line in f:
			if line[0] != '%':
				word = unicode(line.strip(), 'utf-8')
				#print "Adding %r (%s)" % (word, word)
				mtree.add(word)
				loaded_words += 1
				if loaded_words >= words_limit:
					break
				if loaded_words % 100 == 0:
					print '\r%d words indexed' % loaded_words, ; sys.stdout.flush()
	print '\r%d words indexed' % loaded_words
	times = t.getTimes()
	print 'TIMES: %0.2fuser %0.2ftotal' % times
	print 
	
	while True:
		word = unicode(raw_input("Type a word: "), 'utf-8')
		t = Timer()
		for near in mtree.get_nearest(word, limit=10):
			print '\t%d %s' % (near.distance, near.data)
		times = t.getTimes()
		print 'TIMES: %0.2fuser %0.2ftotal' % times
		print



if __name__ == '__main__':
	main()
